sdks/python/apache_beam/runners/portability/fn_api_runner.py

```diff
@@ -1480,42 +1664,54 @@ def process_bundle(self, inputs, expected_outputs):
 
     # Register the bundle descriptor, if needed - noop if already registered.
     registration_future = self._register_bundle_descriptor()
+    # Check that the bundle was successfully registered.
+    if registration_future and registration_future.get().error:
+      raise RuntimeError(registration_future.get().error)
 
     split_manager = self._select_split_manager()
-    if not split_manager:
-      # If there is no split_manager, write all input data to the channel.
+    if not split_manager and isinstance(self._worker_handler,
+                                        EmbeddedWorkerHandler):
+      # If there is no split_manager, write all input data to the channel. For
+      # EmbeddedWorkerHandler, send data first, then send request.
       for transform_id, elements in inputs.items():
         self._send_input_to_worker(
             process_bundle_id, transform_id, elements)
 
-    # Check that the bundle was successfully registered.
-    if registration_future and registration_future.get().error:
-      raise RuntimeError(registration_future.get().error)
-
     # Actually start the bundle.
     process_bundle_req = beam_fn_api_pb2.InstructionRequest(
         instruction_id=process_bundle_id,
         process_bundle=beam_fn_api_pb2.ProcessBundleRequest(
             process_bundle_descriptor_reference=self._bundle_descriptor.id))
-    result_future = self._controller.control_handler.push(process_bundle_req)
+    result_future = self._worker_handler.control_handler.push(
+        process_bundle_req, self._worker_handler._worker_id)
+
+    if not split_manager and isinstance(self._worker_handler,
```

Comment:
What I observed here is, with non-GRPC worker handler, if we send request before sending data, bundle is treated as empty and processed before data is sent so it always return nothing. With GRPC worker handlers, request should be sent before data, because we need to multiplexing data according to the request. That's why I split it into two parts, but I don't think this is ideal way to handle this case.

Reply:
Using the ideas above, we should avoid the need to send the request before the data in the GRPC case. (It is true that split_manager is incompatible with EmbeddedWorkerHandler, wouldn't hurt to assert this.)

(Nit: it's not clear that `isinstance(GrpcWorkerHandler)` and `isinstance(EmbeddedWorkerHandler)` cover both cases, e.g. someone could add a third type without knowing they had to fix the logic here. If we continued with this, it'd be better to do `isinstance(GrpcWorkerHandler)` vs. `not isinstance(GrpcWorkerHandler)`)